Fix cleaning of the modules

bvdmitri opened the pull request:
This PR attempts to fix #2673 . The problem is that the new cleaning functionality aggressively tries to assign `nothing` to all names defined in the module. It does check against `const`, but this isn't enough since users can write typed assignments, e.g. `a::String = "b"`. This would later cause problems when assigning `a` to `nothing`. This PR fixes it by wrapping the `@eval` in `try-catch` block and also adding this as an example in the tests.

fredrikekre commented on src/expander_pipeline.jl:
```diff
@@ -6,7 +6,13 @@ function clear_module!(M::Module)
+            # still attempt to set it, but ignore any errors
+            try
+                @eval M $name = $nothing
+            catch _
+            end
```
Perhaps a debug log message could be useful:
```suggestion
            catch err
                @debug "Could not clear variable `$name` by assigning `nothing`" err
            end
```

fredrikekre reviewed (approved):
If you update the version number in Project.toml we can make a quick release I think.

bvdmitri commented:
So the tests failed on Julia 1.6 since there it was not allowed to use typed declarations for global variables. I changed the example block to use `@static` check. I've also updated the CHANGELOG.md and bumped the version

fingolfin commented on CHANGELOG.md:
```diff
@@ -3,6 +3,12 @@
+
+## Fixed
+
+* Fixed a bug in the cleaning of `@example` blocks with `a::SomeType = somevalue` typed declarations. ([#2673, #2674])
```
Quick unimportant side-question to maintainers: is it usual to include the issue *and* PR here? I don't mind either way, but in my recent PRs I didn't do it, and wonder if I should have.

bvdmitri replied:
for consistency with other changes I only linked PR

bvdmitri replied:
Something is that I still don't understand is that sometimes it is using a single `#` or double `##` to define a new version and it doesn't seem consistent, e.g.
```
# Version [v1.10.0] - 2025-03-31

...

## Version [v1.9.0] - 2025-03-17
``` 

I think it used to use `##` everywhere, but `1.10` didn't do that (accidentally) ? I leave the decision to maintainers. Thanks for reviewing my PR!

asinghvi17 replied:
I usually see only PR numbers, or "Made changes X, Y and Z to fix issue #123 [#124]", where 123 is the issue and 124 is a PR.  But they are generally not together.

fingolfin commented on test/clear_module/src/index.md:
````diff
@@ -21,3 +21,33 @@ a = MyMutableStruct(1)
+
+```@example example-3
+nothing() = 3 # it is possible to use `nothing` as a function name
+a = nothing()
````
Same question

bvdmitri replied:
Same answer as above :)

fingolfin commented on test/clear_module/src/index.md:
````diff
@@ -21,3 +21,33 @@ a = MyMutableStruct(1)
+
+```@example example-4
+@static if VERSION >= v"1.8"
+typed::String = "string" # it is possible to use `::` to specify the type of a variable
````
Can we detect this (and how) to properly fix the issue (in addition to also having the try/catch)

fingolfin reviewed (approved):
Thank you

bvdmitri commented on test/clear_module/src/index.md:
````diff
@@ -21,3 +21,33 @@ a = MyMutableStruct(1)
+```
+
+```@example example-2
+nothing = 1 # it is possible to use `nothing` as a variable name
````
It didn't cause an issue, but when trying to reduce MWE I was experimenting with different "weird" Julia expressions which might have potentially be the cause of the issue. At the end it had nothing to do with defining `nothing` but I put it as a safeguard against future regressions, but I agree with you that `$nothing` should already prevent that because it should refer to `Base.nothing`

fingolfin replied:
OK Thanks! And I agree it is a good idea to out this test in

asinghvi17 commented on src/expander_pipeline.jl:
```diff
@@ -6,7 +6,14 @@ function clear_module!(M::Module)
+            try
+                @eval M $name = $nothing
+            catch err
+                @debug "Could not clear variable `$name` by assigning `nothing`" err
```
```suggestion
                @debug "Could not clear variable `$name` from module `$M` by assigning `nothing`" err

asinghvi17 commented on CHANGELOG.md:
Seems I lost the commit message so commenting here - the Keep a Changelog format and previous Documenter versions all prescribe this format (h2 for versions, h3 for categories within versions).

mortenpi commented:
Thank you! Release here: https://github.com/JuliaRegistries/General/pull/128124
